ts/routes/graphs/true-retention.ts

```diff
@@ -1,103 +1,116 @@
 // Copyright: Ankitects Pty Ltd and contributors
 // License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
-import type { GraphsResponse } from "@generated/anki/stats_pb";
-import * as tr from "@generated/ftl";
+import * as t9n from "@generated/ftl";
 import { localizedNumber } from "@tslib/i18n";
 import { RevlogRange } from "./graph-helpers";
 
-interface TrueRetentionData {
+export interface TrueRetentionData {
     youngPassed: number;
     youngFailed: number;
     maturePassed: number;
     matureFailed: number;
 }
 
-function calculateRetention(passed: number, failed: number): string {
-    const total = passed + failed;
-    if (total === 0) {
-        return "0%";
-    }
-    return localizedNumber((passed / total) * 100, 1) + "%";
+export interface PeriodTrueRetentionData {
+    today: TrueRetentionData;
+    yesterday: TrueRetentionData;
+    week: TrueRetentionData;
+    month: TrueRetentionData;
+    year: TrueRetentionData;
+    allTime: TrueRetentionData;
+}
+
+export enum DisplayMode {
+    Young,
+    Mature,
+    All,
+    Summary,
 }
 
-enum Scope {
+export enum Scope {
     Young,
     Mature,
-    Total,
+    All,
 }
 
-function createStatsRow(period: string, data: TrueRetentionData, scope: Scope): string {
-    let pass: number, fail: number, retention: string;
+export function getPassed(data: TrueRetentionData, scope: Scope): number {
     switch (scope) {
         case Scope.Young:
-            pass = data.youngPassed;
-            fail = data.youngFailed;
-            retention = calculateRetention(data.youngPassed, data.youngFailed);
-            break;
+            return data.youngPassed;
         case Scope.Mature:
-            pass = data.maturePassed;
-            fail = data.matureFailed;
-            retention = calculateRetention(data.maturePassed, data.matureFailed);
-            break;
-        case Scope.Total:
-            pass = data.youngPassed + data.maturePassed;
-            fail = data.youngFailed + data.matureFailed;
-            retention = calculateRetention(pass, fail);
-            break;
+            return data.maturePassed;
+        case Scope.All:
+            return data.youngPassed + data.maturePassed;
+        default:
+            throw new Error(`Unexpected scope: ${scope}`);
```

Comment:
There are a few instances of this pattern in your changes. The problem with doing it this way is that if a new enum variant is added in the future, the linter won't alert us that we haven't handled it. Better to make use of an existing tool we have (which surprisingly, is not currently used anywhere else in the codebase):

```suggestion
            assertUnreachable(scope);
```